src/Model/Table/SqlTraceTrait.php

```diff
@@ -0,0 +1,54 @@
+<?php
+/**
+ * Copyright (c) Cake Software Foundation, Inc. (http://cakefoundation.org)
+ *
+ * Licensed under The MIT License
+ * For full copyright and license information, please see the LICENSE.txt
+ * Redistributions of files must retain the above copyright notice.
+ *
+ * @copyright     Copyright (c) Cake Software Foundation, Inc. (http://cakefoundation.org)
+ * @link          http://cakephp.org CakePHP(tm) Project
+ * @since         3.11.5
+ * @license       http://www.opensource.org/licenses/mit-license.php MIT License
+ */
+namespace DebugKit\Model\Table;
+
+use Cake\Core\Configure;
+use Cake\Error\Debugger;
+
+/**
+ * Add this trait to your Table class to append the file reference of where a Query object was created.
+ *
+ * @mixin \Cake\ORM\Table
+ */
+trait SqlTraceTrait
+{
+    /**
+     * Creates a new Query instance for this repository
+     *
+     * @return \Cake\ORM\Query
+     */
+    public function query()
+    {
+        $query = parent::query();
+
+        if (!Configure::read('debug') || $query->clause('epilog') !== null) {
+            return $query;
+        }
+
+        $traces = Debugger::trace(['start' => 2, 'depth' => 3, 'format' => 'array']);
+        $file = 'n/a';
+        $line = 0;
+
+        foreach ($traces as $trace) {
+            $path = $trace['file'];
+            $line = $trace['line'];
+            $file = Debugger::trimPath($path);
+            if (defined('CAKE_CORE_INCLUDE_PATH') && strpos($path, CAKE_CORE_INCLUDE_PATH) !== 0) {
+                break;
+            }
+        }
+
+        return $query->epilog($query->newExpr(sprintf('/* %s (line %s) */', $file, $line)));
```

Comment:
I personally use `modifier()` instead of epilog for the same usage. `modifier` also has the advantage that it  automatically concatenates. It looks like

```sql
SELECT /* ... */
field1, field2...
```

Having it at the top is also great when you look at the currently running queries directly in your database

Reply:
Thanks, I'll give it a try. I like the comment at the bottom (feels out of the way), but I'll give modifier a try (auto concating is worry free).